src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs

```diff
@@ -0,0 +1,790 @@
+        }
+
+        private void WriteCompletionMessage(CompletionMessage message, ref Utf8JsonWriter writer)
+        {
+            WriteInvocationId(message, ref writer);
+            if (!string.IsNullOrEmpty(message.Error))
+            {
+                writer.WriteString(ErrorPropertyNameBytes, message.Error);
+            }
+            else if (message.HasResult)
+            {
+                var bytes = JsonSerializer.ToBytes(message.Result, message.Result?.GetType());
+                using var token = JsonDocument.Parse(bytes);
+                token.RootElement.WriteAsProperty(ResultPropertyNameBytes, ref writer);
+            }
+        }
+
+        private void WriteCancelInvocationMessage(CancelInvocationMessage message, ref Utf8JsonWriter writer)
+        {
+            WriteInvocationId(message, ref writer);
+        }
+
+        private void WriteStreamItemMessage(StreamItemMessage message, ref Utf8JsonWriter writer)
+        {
+            WriteInvocationId(message, ref writer);
+
+            var bytes = JsonSerializer.ToBytes(message.Item, message.Item?.GetType());
+            using var token = JsonDocument.Parse(bytes);
+            token.RootElement.WriteAsProperty(ItemPropertyNameBytes, ref writer);
+        }
+
+        private void WriteInvocationMessage(InvocationMessage message, ref Utf8JsonWriter writer)
+        {
+            WriteInvocationId(message, ref writer);
+            writer.WriteString(TargetPropertyNameBytes, message.Target);
+
+            WriteArguments(message.Arguments, ref writer);
+
+            WriteStreamIds(message.StreamIds, ref writer);
+        }
+
+        private void WriteStreamInvocationMessage(StreamInvocationMessage message, ref Utf8JsonWriter writer)
+        {
+            WriteInvocationId(message, ref writer);
+            writer.WriteString(TargetPropertyNameBytes, message.Target);
+
+            WriteArguments(message.Arguments, ref writer);
+
+            WriteStreamIds(message.StreamIds, ref writer);
+        }
+
+        private void WriteCloseMessage(CloseMessage message, ref Utf8JsonWriter writer)
+        {
+            if (message.Error != null)
+            {
+                writer.WriteString(ErrorPropertyNameBytes, message.Error);
+            }
+        }
+
+        private void WriteArguments(object[] arguments, ref Utf8JsonWriter writer)
+        {
+            writer.WriteStartArray(ArgumentsPropertyNameBytes);
+            foreach (var argument in arguments)
+            {
+                var type = argument?.GetType();
+                if (type == typeof(DateTime))
+                {
+                    writer.WriteStringValue((DateTime)argument);
+                }
+                else if (type == typeof(DateTimeOffset))
+                {
+                    writer.WriteStringValue((DateTimeOffset)argument);
+                }
+                else
+                {
+                    var bytes = JsonSerializer.ToBytes(argument, type);
```

Comment:
wut

Reply:
There is no way to serialize into a writer, so we need to get the bytes of the object, then create a JsonElement and write the JsonElement into the writer.

Reply:
Raise an issue on corefx. You should be able to do this.